courses/models.py

```diff
@@ -524,3 +526,164 @@ class ProgramEnrollmentAudit(AuditModel):
+    """
+    # Set a custom manager for this model. This lets us do a couple useful things:
+    # (1) Apply commonly-needed filters by a short, easily-recognized name, and (2) compose a query piece-by-piece.
+    # Docs: https://docs.djangoproject.com/en/1.10/topics/db/managers/#creating-a-manager-with-queryset-methods
+    objects = FinalGradeQuerySet.as_manager()
+
+    user = models.ForeignKey(User, null=False, on_delete=models.CASCADE)
+    course_run = models.ForeignKey(CourseRun, null=False, on_delete=models.CASCADE)
+    grade = models.FloatField(
+        null=True,
+        validators=[MinValueValidator(0.0), MaxValueValidator(1.0)]
+    )
+    passed = models.BooleanField(default=False)
+    status = models.CharField(
+        null=False,
+        choices=[(status, status) for status in FinalGradeStatus.ALL_STATUSES],
+        default=FinalGradeStatus.PENDING,
+        max_length=30,
+    )
+    course_run_paid_on_edx = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = ('user', 'course_run')
+
+    @classmethod
+    def get_audit_class(cls):
+        return FinalGradeAudit
+
+    def to_dict(self):
+        return serialize_model_object(self)
+
+    @property
+    def grade_percent(self):
+        """Returns the grade field value as a number out of 100 (or None if the value is None)"""
+        return self.grade * 100 if self.grade is not None else None
+
+    @classmethod
+    def get_frozen_users(cls, course_run):
+        """
+        Returns all the users with a final grade in a given course run
+        """
+        return list(
+            cls.objects.filter(
+                course_run=course_run, status=FinalGradeStatus.COMPLETE
+            ).values_list('user', flat=True)
+        )
+
+    def __str__(self):
+        return 'Grade in course "{course_id}", user "{user}", value {grade}'.format(
+            user=self.user.username,
+            grade=self.grade,
+            course_id=self.course_run.edx_course_key
+        )
+
+    def save(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        """Overridden method to run validation"""
+        self.full_clean()
+        return super().save(*args, **kwargs)
+
+
+class FinalGradeAudit(AuditModel):
+    """
+    Audit table for the Final Grade
+    """
+    final_grade = models.ForeignKey(FinalGrade, null=True, on_delete=models.SET_NULL)
+
+    @classmethod
+    def get_related_field_name(cls):
+        return 'final_grade'
+
+    def __str__(self):
+        return 'Grade audit for user "{user}", course "{course_id}"'.format(
+            user=self.final_grade.user,
+            course_id=self.final_grade.course_run.edx_course_key
+        )
+
+
+class CourseCertificate(TimestampedModel):
+    """
+    Model for storing course certificates
+    """
+    user = models.ForeignKey(User, models.SET_NULL, null=True, related_name='course_certificates')
+    course = models.ForeignKey(Course, models.SET_NULL, null=True)
```

Comment:
How is the course duration supposed to be calculated? I.e.say, a learner passes `SEED Digital Learning 100`, the certificate also displays duration of the course. How do we get that information? If we simply look for a `FinalGrade` for the user for `SEED Digital Learning 100` that was a passing grade, aren't we now leaving room open for errors? The learner could perhaps revisit the course to improve their grade in the future, or even enroll in another program that includes the same course, in that case we will have multiple `FinalGrade` objects for the same user for the same course. This also applies to the `ProgramCertificate` model.

Reply:
This is something that's a little different between MicroMasters and xPRO. In xPRO, the certificate in xPRO should have a foreign key to CoursseRun instead of Course. The certificate will be associated with a particular run of the course (unlike MicroMasters)

Reply:
That makes sense, should take care of the duration issue as well.

Reply:
oh yes. I forgot about duration over certificates.